core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1263,21 +1445,205 @@ class KafkaApis(val requestChannel: RequestChannel,
       )
 
       // call the share partition manager to fetch messages from the local replica
-      sharePartitionManager.fetchMessages(
-        groupId,
-        shareFetchRequest.data.memberId,
-        params,
-        interesting.asJava
-      ).whenComplete { (responsePartitionData, throwable) =>
-        if (throwable != null) {
+      try{
+        val responsePartitionData = sharePartitionManager.fetchMessages(
+          groupId,
+          shareFetchRequest.data.memberId,
+          params,
+          interesting.asJava
+        ).get()
+        processResponseCallback(responsePartitionData.asScala.toMap)
+      } catch {
+        case e: Throwable => throw e
+      }
+    }
+  }
+
+  /**
+   * Handle a shareFetch request
+   */
+  def handleShareFetchRequest(request: RequestChannel.Request): Unit = {
+    val shareFetchRequest = request.body[ShareFetchRequest]
+
+    if (!config.isShareGroupEnabled) {
+      // The API is not supported when the configuration `group.share.enable` has not been set
+      requestHelper.sendMaybeThrottle(request, shareFetchRequest.getErrorResponse(Errors.UNSUPPORTED_VERSION.exception))
+      CompletableFuture.completedFuture[Unit](())
+      return
+    }
+    val topicNames = metadataCache.topicIdsToNames()
+    val sharePartitionManager : SharePartitionManager = sharePartitionManagerOption match {
+      case Some(manager) => manager
+      case None => throw new IllegalStateException("ShareFetchRequest received but SharePartitionManager is not initialized")
+    }
+
+    val groupId = shareFetchRequest.data.groupId
+    val clientId = request.header.clientId
+    val memberId = shareFetchRequest.data().memberId()
+    val shareSessionEpoch = shareFetchRequest.data().shareSessionEpoch()
+
+    val shareFetchData = shareFetchRequest.shareFetchData(topicNames)
+    val forgottenTopics = shareFetchRequest.forgottenTopics(topicNames)
+
+    // TODO : replace this initialization when share fetch session metadata is sent by the client in the shareFetchRequest
+
+    def isAcknowledgeDataPresentInFetchRequest() : Boolean = {
+      var isAcknowledgeDataPresent = false
+      shareFetchRequest.data().topics().forEach ( topic => {
+        breakable{
+          topic.partitions().forEach ( partition => {
+            if (partition.acknowledgementBatches() != null && !partition.acknowledgementBatches().isEmpty) {
+              isAcknowledgeDataPresent = true
+              break()
+            } else {
+              isAcknowledgeDataPresent = false
+            }
+          })
+        }
+      })
+      isAcknowledgeDataPresent
+    }
+
+    var shareAcknowledgeResponse : ShareAcknowledgeResponse = null
+    var shareFetchResponse : ShareFetchResponse = null
+
+    // Handling the Acknowledgements from the ShareFetchRequest
+    val shareSessionError: Errors = sharePartitionManager.acknowledgeShareSessionCacheUpdate(groupId, Uuid.fromString(memberId), shareSessionEpoch)
```

Comment:
This piece seems wrong. When sending the first request in a session, you would expect no acknowledge data in the the fetch request, but you would also expect the share session would not be in the cache yet. So, we get an error code for the acknowledgements in this case, when I would say it should be Errors.NONE.

Reply:
Thanks for the review, I will make this change in the next commit